src/Shared.TextureContainer/WebP/Texture2DExt.cs

```diff
@@ -0,0 +1,281 @@
+
+			byte[] lRawData = LoadRGBAFromWebP(lData, ref lWidth, ref lHeight, lMipmaps, out lError, scalingFunction);
+
+            if (lError == Error.Success)
+            {
+                lTexture2D = new Texture2D(lWidth, lHeight, TextureFormat.RGBA32, lMipmaps, lLinear);
+                lTexture2D.LoadRawTextureData(lRawData);
+                lTexture2D.Apply(lMipmaps, true);
+            }
+
+            return lTexture2D;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="lTexture2D"></param>
+        /// <param name="lData"></param>
+        /// <param name="lError"></param>
+        public static unsafe void LoadWebP(this Texture2D lTexture2D, byte[] lData, out Error lError, ScalingFunction scalingFunction = null)
+        {
+            lError = 0;
+			bool lMipmaps = lTexture2D.mipmapCount != 1;
+
+			int lWidth = 0, lHeight = 0;
+			
+			GetWebPDimensions(lData, out lWidth, out lHeight);
+			
+			byte[] lRawData = LoadRGBAFromWebP(lData, ref lWidth, ref lHeight, lMipmaps, out lError, scalingFunction);
+
+            if (lError == Error.Success)
+            {
+                lTexture2D.LoadRawTextureData(lRawData);
+				lTexture2D.Apply(lMipmaps, true);
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="lTexture2D"></param>
+        /// <param name="lError"></param>
+        /// <returns></returns>
+        public static unsafe byte[] EncodeToWebP(this Texture2D lTexture2D, float lQuality, out Error lError)
+        {
+            lError = 0;
+
+            if (lQuality < -1)  lQuality = -1;
+            if (lQuality > 100) lQuality = 100;
+
+            Color32[] lRawColorData = lTexture2D.GetPixels32();
+            int lWidth  = lTexture2D.width;
+            int lHeight = lTexture2D.height;
+
+            IntPtr lResult = IntPtr.Zero;
+
+            GCHandle lPinnedArray = GCHandle.Alloc(lRawColorData, GCHandleType.Pinned);
+            IntPtr lRawDataPtr = lPinnedArray.AddrOfPinnedObject();
+
+            byte[] lOutputBuffer = null;
+
+            try
+            {
+                int lLength;
+
+                if (lQuality == -1)
+                {
+                    lLength = (int)NativeBindings.WebPEncodeLosslessRGBA(lRawDataPtr, lWidth, lHeight, 4 * lWidth, ref lResult);
+                }
+                else
+                {
+                    lLength = (int)NativeBindings.WebPEncodeRGBA(lRawDataPtr, lWidth, lHeight, 4 * lWidth, lQuality, ref lResult);
+                }
+
+                if (lLength == 0)
+                {
+                    throw new Exception("WebP encode failed!");
+                }
+
+                lOutputBuffer = new byte[lLength];
+                Marshal.Copy(lResult, lOutputBuffer, 0, lLength);
+            }
+            finally
+            {
+                NativeBindings.WebPSafeFree(lResult);
+            }
+
+            lPinnedArray.Free();
```

Comment:
I was looking at it and was a bit curious.
If an exception occurs in try-catch, Free is not called. Is that ok?

Reply:
Probably not, moved it to the finally too. Thanks!

Reply:
I would like to mark this and another case as resolved, but there is no button.
See it as resolved.
![image](https://github.com/IllusionMods/KK_Plugins/assets/4230203/ce82b679-c1fa-4e9c-820c-b78c4d02213b)